### ISSUE TITLE
Indexed `provided` annotation breaks `build` goal (or app startup) in 2.14

### ISSUE THREAD

**Issue by famod** [quarkusio/quarkus#29425]
### Describe the bug

While trying to upgrade from 2.13.3 to 2.14.0/1 I was greeted with a strange `java.lang.ClassNotFoundException: lombok.NonNull` very early in the boot process of the app.

This is in a multi-module project that is working just fine with Quarkus 2.13.3 (and the "old" Jandex 1.2.3).

In a non-app module, a class is using `lombok.NonNull` on a field alongside `lombok.RequiredArgsConstructor`.
`jandex-maven-plugin` is executed in the module to provide an index.
The lombok jar is a `provided` dependency (as recommended generally).
See the reproducer further down.

### Expected behavior

Should work as in 2.13

### Actual behavior

Fails at app startup (with ECJ) or already when building the app via `build` goal (with standard compiler).

### How to Reproduce?

[q_jandex_nonnull.zip](https://github.com/quarkusio/quarkus/files/10071585/q_jandex_nonnull.zip)

`mvn clean verify`:

```
[INFO] --- quarkus-maven-plugin:2.14.1.Final:build (default) @ code-with-quarkus-app ---
[WARNING] [io.quarkus.arc.processor.BeanArchives] Failed to index lombok.NonNull: Class does not exist in ClassLoader QuarkusClassLoader:Deployment Class Loader: PROD@65f470f8
[INFO] [io.quarkus.arc.processor.IndexClassLookupUtils] Class for name: lombok.NonNull was not found in Jandex index. Please ensure the class is part of the index.
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for code-with-quarkus 1.0.0-SNAPSHOT:
[INFO] 
[INFO] code-with-quarkus .................................. SUCCESS [  0.027 s]
[INFO] code-with-quarkus-lib .............................. SUCCESS [  0.924 s]
[INFO] code-with-quarkus-app .............................. FAILURE [  1.511 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.968 s
[INFO] Finished at: 2022-11-22T23:13:14+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:2.14.1.Final:build (default) on project code-with-quarkus-app: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.arc.deployment.ArcProcessor#generateResources threw an exception: java.lang.IllegalStateException: java.util.concurrent.ExecutionException: java.lang.NullPointerException: Annotation class not available: @NonNull
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:918)
[ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:281)
[ERROR] 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:833)
[ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
[ERROR] Caused by: java.util.concurrent.ExecutionException: java.lang.NullPointerException: Annotation class not available: @NonNull
[ERROR] 	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
[ERROR] 	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
[ERROR] 	at io.quarkus.arc.processor.BeanProcessor.generateResources(BeanProcessor.java:319)
[ERROR] 	at io.quarkus.arc.deployment.ArcProcessor.generateResources(ArcProcessor.java:570)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
[ERROR] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
[ERROR] 	... 6 more
[ERROR] Caused by: java.lang.NullPointerException: Annotation class not available: @NonNull
[ERROR] 	at java.base/java.util.Objects.requireNonNull(Objects.java:233)
[ERROR] 	at io.quarkus.arc.processor.AnnotationLiteralProcessor.create(AnnotationLiteralProcessor.java:87)
[ERROR] 	at io.quarkus.arc.processor.BeanGenerator.collectInjectionPointAnnotations(BeanGenerator.java:1937)
[ERROR] 	at io.quarkus.arc.processor.BeanGenerator.wrapCurrentInjectionPoint(BeanGenerator.java:1837)
[ERROR] 	at io.quarkus.arc.processor.BeanGenerator.initConstructor(BeanGenerator.java:704)
[ERROR] 	at io.quarkus.arc.processor.BeanGenerator.createConstructor(BeanGenerator.java:614)
[ERROR] 	at io.quarkus.arc.processor.BeanGenerator.generateClassBean(BeanGenerator.java:355)
[ERROR] 	at io.quarkus.arc.processor.BeanGenerator.generate(BeanGenerator.java:122)
[ERROR] 	at io.quarkus.arc.processor.BeanProcessor$4.call(BeanProcessor.java:258)
[ERROR] 	at io.quarkus.arc.processor.BeanProcessor$4.call(BeanProcessor.java:254)
[ERROR] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[ERROR] 	... 5 more
```

`mvn clean verify -Pquarkus-2.13.4` is fine though (switches from Quarkus 2.14.1 to 2.13.4 and jandex-plugin to the old coordinates and version) 

Please note that this reproducer is not using ECJ (unlike my actual project), hence the error at build time, not boot time.
If required I can provide a ECJ setup like in https://github.com/smallrye/jandex/issues/268.
I actually started with an ECJ setup, stripped it down and was then surprised by the different behavior.

### Output of `java -version`

17.0.5

### Quarkus version or git rev

2.14.1

### Additional information

I _think_ it's an issue in Jandex (cc @Ladicek), but I'm not entirely sure. Might also be related to Arc (cc @mkouba @manovotn )?

If I had to guess (wildly), I'd suppose that Jandex 3 is more strict about missing annotations.

Please note that there have been discussions about why `lombok.NonNull` has `CLASS` retention in the first place: https://github.com/projectlombok/lombok/issues/895
I do think the maintainer has a point though: https://github.com/projectlombok/lombok/issues/895#issuecomment-145653045

**Comment by geoand** [comment]
cc @Ladicek @mkouba

**Comment by Ladicek** [comment]
I think the `AnnotationLiteralProcessor` could and should do 2 things:

1. Ignore class-retained annotation[ instance]s. Annotation literals should only be generated for runtime-retained annotations.
2. Ignore annotation[ instance]s for which the annotation class can't be found. As stated in the Lombok discussion and on many other places, missing annotation classes are perfectly fine.

I'll submit a PR.

**Comment by geoand** [comment]
> missing annotation classes are perfectly fine

+100, this is very easy to forget, but it's very important to realize that it's perfectly valid from a JVM perspective

**Comment by manovotn** [comment]
> Ignore class-retained annotation[ instance]s. Annotation literals should only be generated for runtime-retained annotations.

Agreed.

> Ignore annotation[ instance]s for which the annotation class can't be found. As stated in the Lombok discussion and on many other places, missing annotation classes are perfectly fine.

Not so sure about this - if you ignore a missing CDI annotation (say, a qualifier), that breaks your application but might let you still start it making it very hard to identify the cause.

**Comment by Ladicek** [comment]
Yeah, I'm looking at the code and indeed missing annotation classes for CDI annotations are likely a problem we want to report.

**Comment by famod** [comment]
FTR, this is the boot error when using ECJ:
```
$ java -jar app/target/quarkus-app/quarkus-run.jar 
The DelayedHandler was closed before any children handlers were configured. Messages will be written to stderr.
2022-11-23 10:18:00,223 DEBUG [org.jboss.logging] (main) Logging Provider: org.jboss.logging.JBossLogManagerProvider

Exception in thread "main" java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:61)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:32)
Caused by: java.lang.ExceptionInInitializerError
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.base/java.lang.reflect.ReflectAccess.newInstance(ReflectAccess.java:128)
	at java.base/jdk.internal.reflect.ReflectionFactory.newInstance(ReflectionFactory.java:347)
	at java.base/java.lang.Class.newInstance(Class.java:645)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:69)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:43)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:123)
	at io.quarkus.runner.GeneratedMain.main(Unknown Source)
	... 6 more
Caused by: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.<clinit>(Unknown Source)
	... 17 more
Caused by: java.lang.NoClassDefFoundError: lombok/NonNull
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1012)
	at io.quarkus.bootstrap.runner.RunnerClassLoader.loadClass(RunnerClassLoader.java:105)
	at io.quarkus.bootstrap.runner.RunnerClassLoader.loadClass(RunnerClassLoader.java:65)
	at org.acme.LibBean_Bean.<init>(Unknown Source)
	at io.quarkus.arc.setup.Default_ComponentsProvider.addBeans1(Unknown Source)
	at io.quarkus.arc.setup.Default_ComponentsProvider.getComponents(Unknown Source)
	at io.quarkus.arc.impl.ArcContainerImpl.<init>(ArcContainerImpl.java:117)
	at io.quarkus.arc.Arc.initialize(Arc.java:31)
	at io.quarkus.arc.runtime.ArcRecorder.initContainer(ArcRecorder.java:43)
	at io.quarkus.deployment.steps.ArcProcessor$generateResources844392269.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.ArcProcessor$generateResources844392269.deploy(Unknown Source)
	... 18 more
Caused by: java.lang.ClassNotFoundException: lombok.NonNull
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	at io.quarkus.bootstrap.runner.RunnerClassLoader.loadClass(RunnerClassLoader.java:115)
	at io.quarkus.bootstrap.runner.RunnerClassLoader.loadClass(RunnerClassLoader.java:65)
	... 30 more
```

**Comment by mkouba** [comment]
>> Ignore class-retained annotation[ instance]s. Annotation literals should only be generated for runtime-retained annotations.

> Agreed.

I wonder if we should try to solve this problem more generally because ArC might not be the only user of jandex that got hit by the fact that Jandex 3 does index annotations with `RetentionPolicy.CLASS` (and Jandex 2 does not ;-)...

**Comment by Ladicek** [comment]
So I found that most of the time, `AnnotationLiteralProcessor` is used with CDI annotations (qualifiers and interceptor bindings) and there are just a few places where it can be used with other annotations (when it's fed with all annotations from an injection point). So from ArC perspective, it's probably best if `AnnotationLiteralProcessor` actually not only checks that the annotation class is available, but also that it's runtime-retained -- if it's by accident called with an annotation that doesn't satisfy those conditions, we wanna know. Those few places that deal with arbitrary injection point annotations can be updated to filter the annotations themselves.

@mkouba Do you have anything specific in mind? Plenty people have asked for Jandex to include class-retained annotations, so we can't really move back. ArC unfortunately calls the Jandex API directly on many places, so modifying the `AnnotationStore` to filter out class-retained annotations won't help. I really should start working on the annotation overlay in Jandex, which should allow centralizing many things that are currently done in an ad-hoc fashion...

**Comment by mkouba** [comment]
> Do you have anything specific in mind?

Not really. We could have returned runtime visible annotations only by default and optionally include `RetentionPolicy.CLASS` annotations to make it backward compatible. But it's too late.

> Plenty people have asked for Jandex to include class-retained annotations, so we can't really move back.

I do not say we should move back but it was an incompatible change that may lead to problems like this. And my point is that ArC might not be the only one affected.

**Comment by Ladicek** [comment]
Yeah, agree that the default API could have returned only runtime-retained annotations and class-retained annotations could have been available only through some special-purpose methods. It didn't occur to me back then, unfortunately :disappointed:

It really only covers half of the problem here, though. You can still have runtime-retained annotations whose classes are missing, so that needs fixing anyway.

**Comment by mkouba** [comment]
> It really only covers half of the problem here, though. You can still have runtime-retained annotations whose classes are missing, so that needs fixing anyway.

I agree that a missing annotation is a problem for CDI.

**Comment by Ladicek** [comment]
Weld doesn't have to, because reflection silently filters out annotations whose classes are missing. Since we're processing bytecode, we have to do that ourselves.

**Comment by gsmet** [comment]
> Yeah, agree that the default API could have returned only runtime-retained annotations and class-retained annotations could have been available only through some special-purpose methods. It didn't occur to me back then, unfortunately

Might be an unpopular opinion but given it's a very new feature, if we think it was a bad decision and we want to fix it, it's probably still possible if we document it properly in the announcement. I don't think it's widely used in the wild.
Better than live with it forever.

**Comment by Ladicek** [comment]
> Yes, I'm not talking about "processing" errors but "semantic" errors (i.e. a class is annotated with a qualifier but the qualifier class is not available).

Okay, so what I'm doing is:

1. `AnnotationLiteralProcessor` will still throw if the annotation class is `null`, and most callers don't check that. That's because most callers deal with qualifiers and interceptor bindings, and those must always exist. So if a qualifier class or an interceptor binding class is missing, we'll still fail. There are 2 places that deal with arbitrary annotations on injection points, and those will just check if the annotation class exists or not.
2. This is an extra and I think we wouldn't really have to do it, but I think we want to: `AnnotationLiteralProcessor` will additionally throw if the annotation is class-retained. Again, most callers deal with qualifiers and interceptor bindings, and those must always be runtime-retained. So if a qualifier or interceptor binding is class-retained, we'll fail. I think that's good. The same 2 places mentioned in previous item are modified to just skip class-retained annotations.

WDYT?

> Might be an unpopular opinion but given it's a very new feature, if we think it was a bad decision and we want to fix it, it's probably still possible if we document it properly in the announcement. I don't think it's widely used in the wild.
> Better than live with it forever.

That sounds sensible. I'll see if I can come up with a better API.

**Comment by famod** [comment]
Fix confirmed in 2.14.2! Thanks all, especially @Ladicek .
